glean_parser/translate.py

```diff
@@ -109,6 +110,35 @@ def translate_metrics(
     if util.report_validation_errors(all_objects):
         return 1
 
+    # We transform certain metrics typenames depending on context in the file.
+    # e.g. A `rate` may instead be a `numerator` if its denominator is external.
+    value = all_objects.value
+    counters = {}
+    numerators_by_denominator: Dict[str, Any] = {}
+    for category_key, category_val in value.items():
+        for metric_name, metric in category_val.items():
+            fqmn = category_key + "." + metric_name
+            if getattr(metric, "type", None) == "counter":
+                counters[fqmn] = metric
+            denominator_name = getattr(metric, "denominator_metric", False)
+            if denominator_name:
+                metric.type = "numerator"
+                if denominator_name not in numerators_by_denominator:
+                    numerators_by_denominator[denominator_name] = []
```

Comment:
Could be `numerators_by_denominator.setdefault(denominator_name, [])` to avoid the `if`.